File: src/iac_code/acp/state.py
"""TurnState / ToolCallState — track per-turn and per-tool-call state."""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field


@dataclass
class ToolCallState:
    """Track the streaming state of a single tool call."""

    tool_call_id: str
    tool_name: str
    accumulated_input: str = ""
    title: str = ""
    start_time: float = 0.0

    def __post_init__(self) -> None:
        if self.start_time == 0.0:
            self.start_time = time.monotonic()

    @property
    def elapsed_ms(self) -> int:
        """Milliseconds elapsed since the tool call started."""
        return int((time.monotonic() - self.start_time) * 1000)

    def update_input(self, delta: str) -> None:
        self.accumulated_input += delta
        self._update_title()

    def _update_title(self) -> None:
        """Compute a display title from tool name + streamed arguments."""
        subtitle = _extract_key_argument(self.tool_name, self.accumulated_input)
        title = display_tool_title(self.tool_name)
        self.title = f"{title}: {subtitle}" if subtitle else title


@dataclass
class TurnState:
    """Track all state within a single prompt turn."""

    turn_id: str
    tool_calls: dict[str, ToolCallState] = field(default_factory=dict)

    def start_tool_call(self, tool_call_id: str, tool_name: str) -> ToolCallState:
        state = ToolCallState(tool_call_id=tool_call_id, tool_name=tool_name)
        state.title = display_tool_title(tool_name)
        self.tool_calls[tool_call_id] = state
        return state

    def get_tool_call(self, tool_call_id: str) -> ToolCallState | None:
        return self.tool_calls.get(tool_call_id)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

# Map of tool name -> JSON key whose value makes a good subtitle.
_KEY_ARG_MAP: dict[str, str] = {
    "bash": "command",
    "read_file": "file_path",
    "write_file": "file_path",
    "edit_file": "file_path",
    "glob": "pattern",
    "grep": "pattern",
    "list_files": "path",
    "web_fetch": "url",
}

_SUBTITLE_MAX_LEN = 60


def display_tool_title(tool_name: str) -> str:
    from iac_code.pipeline.display_names import known_tool_display_name

    return known_tool_display_name(tool_name) or tool_name


def _extract_key_argument(tool_name: str, raw_json: str) -> str:
    """Best-effort extraction of a subtitle from partial/complete JSON args."""
    key = _KEY_ARG_MAP.get(tool_name)
    if not key:
        return ""
    try:
        obj = json.loads(raw_json)
        value = obj.get(key, "")
        if isinstance(value, str) and value:
            return value[:_SUBTITLE_MAX_LEN]
    except (json.JSONDecodeError, TypeError, AttributeError):
        # Partial JSON — fall back to naive substring search.
        marker = f'"{key}"'
        idx = raw_json.find(marker)
        if idx == -1:
            return ""
        # Skip past `"key": "` to grab value chars.
        rest = raw_json[idx + len(marker) :]
        rest = rest.lstrip(": ")
        if rest.startswith('"'):
            rest = rest[1:]
            end = rest.find('"')
            snippet = rest[:end] if end != -1 else rest
            return snippet[:_SUBTITLE_MAX_LEN]
    return ""

File: iac-code-rs/crates/iac-code-acp/src/state.rs
use std::collections::BTreeMap;
use std::time::Instant;

use iac_code_protocol::json::JsonValue;

const SUBTITLE_MAX_LEN: usize = 60;

#[derive(Clone, Debug)]
pub struct ToolCallState {
    pub tool_call_id: String,
    pub tool_name: String,
    pub accumulated_input: String,
    pub title: String,
    start_time: Instant,
}

impl ToolCallState {
    pub fn new(tool_call_id: impl Into<String>, tool_name: impl Into<String>) -> Self {
        let tool_name = tool_name.into();
        Self {
            tool_call_id: tool_call_id.into(),
            accumulated_input: String::new(),
            title: tool_name.clone(),
            tool_name,
            start_time: Instant::now(),
        }
    }

    pub fn elapsed_ms(&self) -> u128 {
        self.start_time.elapsed().as_millis()
    }

    pub fn update_input(&mut self, delta: &str) {
        self.accumulated_input.push_str(delta);
        self.update_title();
    }

    fn update_title(&mut self) {
        let subtitle = extract_key_argument(&self.tool_name, &self.accumulated_input);
        self.title = if subtitle.is_empty() {
            self.tool_name.clone()
        } else {
            format!("{}: {}", self.tool_name, subtitle)
        };
    }
}

#[derive(Clone, Debug, Default)]
pub struct TurnState {
    pub turn_id: String,
    pub tool_calls: BTreeMap<String, ToolCallState>,
}

impl TurnState {
    pub fn new(turn_id: impl Into<String>) -> Self {
        Self {
            turn_id: turn_id.into(),
            tool_calls: BTreeMap::new(),
        }
    }

    pub fn start_tool_call(&mut self, tool_call_id: &str, tool_name: &str) -> &ToolCallState {
        self.tool_calls.insert(
            tool_call_id.to_owned(),
            ToolCallState::new(tool_call_id, tool_name),
        );
        self.tool_calls
            .get(tool_call_id)
            .expect("tool call was just inserted")
    }

    pub fn get_tool_call(&self, tool_call_id: &str) -> Option<&ToolCallState> {
        self.tool_calls.get(tool_call_id)
    }

    pub fn get_tool_call_mut(&mut self, tool_call_id: &str) -> Option<&mut ToolCallState> {
        self.tool_calls.get_mut(tool_call_id)
    }
}

pub fn extract_key_argument(tool_name: &str, raw_json: &str) -> String {
    let Some(key) = key_argument(tool_name) else {
        return String::new();
    };
    if let Ok(JsonValue::Object(fields)) = iac_code_protocol::json::parse(raw_json) {
        if let Some(JsonValue::String(value)) = fields.get(key) {
            return truncate_chars(value, SUBTITLE_MAX_LEN);
        }
        return String::new();
    }

    let marker = format!("\"{key}\"");
    let Some(index) = raw_json.find(&marker) else {
        return String::new();
    };
    let rest = raw_json[index + marker.len()..].trim_start_matches([':', ' ']);
    let Some(rest) = rest.strip_prefix('"') else {
        return String::new();
    };
    let end = rest.find('"').unwrap_or(rest.len());
    truncate_chars(&rest[..end], SUBTITLE_MAX_LEN)
}

fn key_argument(tool_name: &str) -> Option<&'static str> {
    match tool_name {
        "bash" => Some("command"),
        "read_file" | "write_file" | "edit_file" => Some("file_path"),
        "glob" | "grep" => Some("pattern"),
        "list_files" => Some("path"),
        "web_fetch" => Some("url"),
        _ => None,
    }
}

fn truncate_chars(value: &str, max_len: usize) -> String {
    value.chars().take(max_len).collect()
}
